address/address.go

```diff
@@ -47,327 +77,236 @@ const (
-		fmt.Fprintf(f, "[%s - %x - %x]", NetworkToString(a.Network()), a.Version(), a.Hash()) // nolint: errcheck
-	case 's':
-		fmt.Fprintf(f, "%s", a.String()) // nolint: errcheck
-	default:
-		fmt.Fprintf(f, "%"+string(c), a.Bytes()) // nolint: errcheck
-	}
+// NewBLSAddress returns an address using the BLS protocol.
+func NewBLSAddress(pubkey []byte) Address {
+	return newAddress(BLS, pubkey)
 }
 
-// MarshalText implements the TextMarshaler interface.
-func (a Address) MarshalText() ([]byte, error) {
-	if a == (Address{}) {
-		return nil, nil
-	}
-
-	return []byte(a.String()), nil
+// NewFromString returns the address represented by the string `addr`.
+func NewFromString(addr string) (Address, error) {
+	return decode(addr)
 }
 
-// UnmarshalText implements the TextUnmarshaler interface.
-func (a *Address) UnmarshalText(in []byte) error {
-	if len(in) == 0 {
-		return nil
-	}
-
-	out, err := NewFromString(string(in))
-	if err != nil {
-		return err
-	}
-
-	copy(a[:], out[:])
-
-	return nil
+// NewFromBytes return the address represented by the bytes `addr`.
+func NewFromBytes(addr []byte) Address {
+	return Address{string(addr)}
 }
 
-func (a Address) String() string {
-	out, err := encode(NetworkToString(a.Network()), a.Version(), a.Hash())
-	if err != nil {
-		// should really not happen
-		panic(err)
-	}
-
-	return out
+// Checksum returns the checksum of `ingest`.
+func Checksum(ingest []byte) []byte {
+	return hash(ingest, checksumHashConfig)
 }
 
-// Bytes returns the byte representation of the address.
-func (a Address) Bytes() []byte {
-	return a[:]
+// ValidateChecksum returns true if the checksum of `ingest` is equal to `expected`>
+func ValidateChecksum(ingest, expect []byte) bool {
+	digest := Checksum(ingest)
+	return bytes.Equal(digest, expect)
 }
 
-// --
-// TODO: find a better place for the things below
+func addressHash(ingest []byte) []byte {
+	return hash(ingest, payloadHashConfig)
+}
 
-// encode encodes hrp(human-readable part) a version(byte) and data(32bit data array)
-func encode(hrp string, version byte, data []byte) (string, error) {
-	if len(hrp) != 2 {
-		return "", fmt.Errorf("hrp has invalid length: hrp=%d", len(hrp))
+func newAddress(protocol Protocol, payload []byte) Address {
+	if protocol < ID || protocol > BLS {
+		panic("invalid protocol")
 	}
-	if len(data) != HashLength {
-		return "", fmt.Errorf("data is malformed: data length=%d", len(data))
+	explen := 1 + len(payload)
+	buf := make([]byte, explen)
+
+	buf[0] = protocol
+	if c := copy(buf[1:], payload); c != len(payload) {
+		panic("copy data length is inconsistent")
 	}
-	for p, c := range hrp {
-		if c < 33 || c > 126 {
-			return "", fmt.Errorf("invalid character human-readable part : hrp[%d]=%d", p, c)
-		}
+	log.Debugf("new address protocol: %x, payload: %v", protocol, payload)
+	return Address{string(buf)}
+}
+
+func encode(network Network, addr Address) (string, error) {
+	// TODO(frrist): not sure if this is the correct thing to do here
```

Comment:
Don't merge with this TODO still here. I suggest in general using a different tag like DONOTMERGE or similar to help you and reviewers address questions that you ask via the code.